Wikipedia/Code/ArticlesCollectionViewController.swift

```diff
@@ -0,0 +1,487 @@
+        let recentlyAddedSortAction = SortActionType.byRecentlyAdded.action(with: [NSSortDescriptor(keyPath: \ReadingListEntry.createdDate, ascending: false)], handler: handler)
+        
+        return [titleSortAction.type: titleSortAction, recentlyAddedSortAction.type: recentlyAddedSortAction]
+    }()
+    
+    lazy var sortAlert: UIAlertController = {
+        alert(title: CommonStrings.sortAlertTitle, message: nil)
+    }()
+    
+    func article(at indexPath: IndexPath) -> WMFArticle? {
+        guard let fetchedResultsController = fetchedResultsController, fetchedResultsController.isValidIndexPath(indexPath) else {
+            return nil
+        }
+        let entry = fetchedResultsController.object(at: indexPath)
+        
+        guard let article = entry.articleKey.flatMap(dataStore.fetchArticle(withKey:)) else {
+            return nil
+        }
+        return article
+    }
+    
+    func article(for entry: ReadingListEntry) -> WMFArticle? {
+        return entry.articleKey.flatMap(dataStore.fetchArticle(withKey: ))
+    }
+    
+    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        super.scrollViewDidScroll(scrollView)
+        editController.transformBatchEditPaneOnScroll()
+    }
+    
+    func articleURL(at indexPath: IndexPath) -> URL? {
+        return article(at: indexPath)?.url
+    }
+    
+    func readingLists(for article: WMFArticle) -> [ReadingList] {
+        guard let moc = article.managedObjectContext else {
+            return []
+        }
+        
+        let request: NSFetchRequest<ReadingList> = ReadingList.fetchRequest()
+        request.predicate = NSPredicate(format: "ANY articles == %@ && isDefault == NO", article)
+        request.sortDescriptors = [NSSortDescriptor(keyPath: \ReadingList.canonicalName, ascending: true)]
+        
+        do {
+            return try moc.fetch(request)
+        } catch {
+            DDLogError("Error fetching lists: \(error)")
+            return []
+        }
+    }
+    
+    // MARK: - ColumnarCollectionViewLayoutDelegate
+    
+    override func collectionView(_ collectionView: UICollectionView, estimatedHeightForItemAt indexPath: IndexPath, forColumnWidth columnWidth: CGFloat) -> ColumnarCollectionViewLayoutHeightEstimate {
+        // The layout estimate can be re-used in this case becuause both labels are one line, meaning the cell
+        // size only varies with font size. The layout estimate is nil'd when the font size changes on trait collection change
+        if let estimate = cellLayoutEstimate {
+            return estimate
+        }
+        var estimate = ColumnarCollectionViewLayoutHeightEstimate(precalculated: false, height: 60)
+        guard let placeholderCell = layoutManager.placeholder(forCellWithReuseIdentifier: reuseIdentifier) as? SavedArticlesCollectionViewCell, let entry = fetchedResultsController?.object(at: indexPath) else {
```

Comment:
`let entry = fetchedResultsController?.object(at: indexPath)` previously this was an `entry(at: IndexPath)` function that ensured `indexPath` was valid before requesting it from the `fetchedResultsController` in an ideal world, there's never an invalid index path due to updates from the FRC being handled properly, but in practice it's nice to have the extra insurance against a crash

Reply:
This comment applies to the other `let entry = fetchedResultsController?.object(at: indexPath)` uses in the file as well